### PR TITLE
StashCache tests (SOFTWARE-3360)

### PR DESCRIPTION
**Pull request by matyasselmeci**
Happy results: http://vdt.cs.wisc.edu/tests/20181207-1548/results.html

### CODE REVIEW

File: osgtest/tests/test_150_xrootd.py [hunk 1]
```diff
@@ -58,7 +63,7 @@ def test_01_start_xrootd(self):
                             owner="xrootd",
                             chown=(user.pw_uid, user.pw_gid))
 
-            files.append(core.config['xrootd.config'], XROOTD_CFG_TEXT % sec_protocol, owner='xrootd', backup=True)
+            files.append(core.config['xrootd.config'], XROOTD_CFG_TEXT % (sec_protocol, core.config['xrootd.port']), owner='xrootd', backup=True)
```
Comment (brianhlin) [inline osgtest/tests/test_150_xrootd.py]: Can we split this up so it fits within the 120 char limit?

Comment (matyasselmeci) [reply osgtest/tests/test_150_xrootd.py]: Done

File: osgtest/tests/test_155_stashcache.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+xrd.trace all
+cms.trace all
+""".format(**globals())
+
+
+CACHES_JSON_TEXT = """\
+[
+{"name":"root://localhost", "status":1, "longitude":-89.4012, "latitude":43.0731}
+]
+"""
```
Comment (brianhlin) [inline osgtest/tests/test_155_stashcache.py]: Wow this is a lot of config. Can we simplify this after https://github.com/opensciencegrid/StashCache-Daemon/pull/21 gets released?

File: osgtest/tests/test_155_stashcache.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            ("origin_xroot_port", ORIGIN_XROOT_PORT),
+            ("cache_xroot_port", CACHE_XROOT_PORT)
+        ]:
+            _setcfg(key, val)
+
+        xrootd_user = pwd.getpwnam("xrootd")
+        for d in [_getcfg("origin_dir"), _getcfg("cache_dir"),
+                  os.path.dirname(_getcfg("caches_json_path"))]:
+            files.safe_makedirs(d)
+            os.chown(d, xrootd_user.pw_uid, xrootd_user.pw_gid)
```
Comment (brianhlin) [inline osgtest/tests/test_155_stashcache.py]: What do you think about making `files.safe_makedirs` take optional args for chown/chmod (defaulting to `root:root` and `0644`)?

File: osgtest/tests/test_460_stashcache.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        ("testfile%d" % x, str(random.random()))
+        for x in range(4)
+    ]
+
+    def assertCached(self, name, contents):
+        fpath = os.path.join(_getcfg("cache_dir"), name)
+        self.assertTrue(os.path.exists(fpath),
+                        name + " not cached")
+        self.assertEqual(contents, files.read(fpath, as_single_string=True),
+                         "cached file %s contents do not match expected" % name)
```
Comment (brianhlin) [inline osgtest/tests/test_460_stashcache.py]: Should we include the actual content vs the expected content in the message?

File: osgtest/tests/test_460_stashcache.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                        contents, backup=False, chmod=0o644,
+                        chown=(xrootd_user.pw_uid, xrootd_user.pw_gid))
+
+    def test_02_xroot_fetch_from_origin(self):
+        name, contents = self.testfiles[0]
+        result, _, _ = \
+            core.check_system(["xrdcp", "-d1", "-N", "-f",
+                               "root://localhost:%d//%s" % (_getcfg("origin_xroot_port"), name),
+                               "-"], "Checking xroot copy from origin")
+        self.assertEqual(result, contents, "downloaded file does not match expected")
```
Comment (brianhlin) [inline osgtest/tests/test_460_stashcache.py]: Can we test HTTP fetches from the origin or is that not supported?

File: osgtest/tests/test_155_stashcache.py (hunk 2 above)
Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: Slightly. If we start with the cache server config from that PR (and it doesn't look like it has any origin server config changes), I'd need to make the following config changes as part of setup:

- set `cachedir` and `sitename` in `10-osg-site-local.cfg`
- change the HTTPS port in `40-osg-http.cfg`
- change the HTTP port in `40-osg-http.cfg`
- delete `40-osg-monitoring.cfg`
- change both `pss.origin` settings in `40-osg-xcache.cfg`

So basically we'd be looking at a bunch of `sed`-equivalents instead of one config file write. I suppose that's probably the right thing to do in the long run anyway...

File: osgtest/tests/test_155_stashcache.py (hunk 3 above)
Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: I'd rather have it follow the semantics of its namesake (`os.makedirs`) as opposed to making something different and unexpected.

But you just reminded me that the chown is redundant anyway since osg-test always runs as root.

File: osgtest/tests/test_460_stashcache.py (hunk 5 above)
Comment (matyasselmeci) [reply osgtest/tests/test_460_stashcache.py]: The origin isn't running the HTTP plugin.

File: osgtest/tests/test_460_stashcache.py (hunk 4 above)
Comment (matyasselmeci) [reply osgtest/tests/test_460_stashcache.py]: Sure. I added a new `assertEqualVerbose` method that prints the expected and actual values on failure.

File: osgtest/tests/test_155_stashcache.py (hunk 3 above)
Comment (brianhlin) [reply osgtest/tests/test_155_stashcache.py]: Maybe I'm misunderstanding your last comment but the chown that you removed is setting ownership to `xrootd:xrootd` and now it's chowning the dirs to `root:root`

File: osgtest/tests/test_155_stashcache.py (hunk 2 above)
Comment (brianhlin) [reply osgtest/tests/test_155_stashcache.py]: Would we be able to instead write a `99-local.cfg` that overrides the values in the files you mentioned or does XRootD config not work that way?

File: osgtest/tests/test_155_stashcache.py (hunk 3 above)
Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: /me looks at code again
I am a moron. Fixed.

File: osgtest/tests/test_155_stashcache.py (hunk 2 above)
Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: That's a good question. @bbockelm, if there are duplicate settings, do the later lines override the earlier ones?

**Comment by matyasselmeci** [comment]
More happy tests: http://vdt.cs.wisc.edu/tests/20181211-0842/packages.html

File: osgtest/tests/test_155_stashcache.py (hunk 2 above)
Comment (bbockelm) [reply osgtest/tests/test_155_stashcache.py]: Depends on the configuration variable.

Xrootd configs are not a hash lookup (as with an INI file or HTCondor config) but rather linear parsing.  So, some attributes are additive (IIRC, the `all.export` lines are an example of this) where, with others, the last-specified wins.

I think everything you mention above is "last-specified wins".

Not clear - why does `40-osg-monitoring.cfg` need to be deleted though?  For the unit tests, I'd be really happy if we can stick to changing the configs in ways that we recommend to admins...

Comment (bbockelm) [reply osgtest/tests/test_155_stashcache.py]: In that case, can you add a `-itb` suffix to the monitoring hostnames?  I don't think there's something listening on UDP there right now, but I'd like to have these show up eventually (as opposed to dropping them...)

Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: Be careful what you wish for :P

So the new tasks will be:

- set `cachedir` and `sitename` in `10-osg-site-local.cfg`
- change the HTTPS port in `40-osg-http.cfg`
- change the HTTP port in `40-osg-http.cfg`
- replace xrd-*.osgstorage.org with xrd-*-itb.osgstorage.org in `40-osg-monitoring.cfg`
- change both `pss.origin` settings in `40-osg-xcache.cfg`

Comment (bbockelm) [reply osgtest/tests/test_155_stashcache.py]: Only the first bullet is correct - for the others, we should be able to do this solely via a 99-overrides.cfg

Comment (bbockelm) [reply osgtest/tests/test_155_stashcache.py]: Only the first bullet is correct - for the others, we should be able to do this solely via a 99-overrides.cfg

Comment (matyasselmeci) [reply osgtest/tests/test_155_stashcache.py]: OK. Well, we'll do that for the next version of the tests.
